lib/Dialect/Random/IR/RandomOps.td

```diff
@@ -0,0 +1,105 @@
+#ifndef LIB_DIALECT_RANDOM_IR_RANDOMOPS_TD_
+#define LIB_DIALECT_RANDOM_IR_RANDOMOPS_TD_
+
+include "lib/Dialect/Random/IR/RandomDialect.td"
+include "lib/Dialect/Random/IR/RandomTypes.td"
+include "mlir/IR/BuiltinAttributes.td"
+include "mlir/IR/CommonAttrConstraints.td"
+include "mlir/IR/CommonTypeConstraints.td"
+include "mlir/IR/OpBase.td"
+include "mlir/Interfaces/InferTypeOpInterface.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+
+class Random_Op<string mnemonic, list<Trait> traits = []> :
+        Op<Random_Dialect, mnemonic, traits> {
+  let cppNamespace = "::mlir::heir::random";
+  let assemblyFormat = "operands attr-dict `:` functional-type(operands, results)";
+}
+
+def Random_InitOp: Random_Op<"init_prng", [Pure]> {
+  let summary = "Initializes the pseudorandom number generator with a seed.";
+  let description = [{
+    Initializes the PRNG with a seed. The seed is dynamically provided due to
+    protocols that agree on shared randomness. The PRNG is used to initialized
+    the random distributions such as the discrete gaussian distribution and the
+    discrete uniform distribution. This initialization also takes as input a
+    number of bits that are generated for each number value sampled (num_bits).
+    For instance, a num_bits of 32 will mean that distributions will generate a
+    32-bit integer value. We expect that the seed initialization is done statically
+    and globally once per thread for all distributions; however, if multiple threads are
+    generating randomness, then seed initialization should be done per thread;
+    otherwise there is no guarantee of consistent behavior. Thread safety is so
+    far not considered.
+  }];
+
+  let arguments = (ins
+    SignlessIntegerLike:$seed,
+    Builtin_IntegerAttr:$num_bits
```

Comment:
I'm OK merging this in as-is, but I don't fully understand why the attribute is needed at all. The `sample` ops are already not constrained to choose types related to this attribute, and I feel like if this were needed, it would make more sense to specify it at sample time.

What's the reason for including it?

Reply:
CC @asraa since I just saw her earlier comment. Why isn't it enough at this level of granularity to let the sample op decide what type of output it wants? The lowering should, IMO, decide how to produce random bits in the end, and AFAIK all random generators boil down to generating a stream of individual random bits.

Reply:
I'll merge it in, and we'll table adjusting the attribute (if needed) to a future PR.

Reply:
> Why isn't it enough at this level of granularity to let the sample op decide what type of output it wants?

A PRNG needs to know the number of bits to produce - like you can specify a 32 bit or 64 bit output from the algorithm.  I don't think that makes sense to infer from the sample op, since a sample producing 4 bits or 8 bits or etc should all be using the same underlying PRNG that may produce a fixed output bit size.

Reply:
For example https://openfhe-development.readthedocs.io/en/stable/api/classlbcrypto_1_1Blake2Engine.html#_CPPv4N8lbcrypto12Blake2EngineE

templates on 32-bits, technically some other underlying data type can be used

Reply:
>  all random generators boil down to generating a stream of individual random bits.

Yeah, this is still true. Maybe you're right that at the abstraction level of random it shouldn't be known, and that at the time the hardware target is known we would have that data chosen then.

Reply:
Yeah I mainly think that it's an implementation detail that should be decided by the lowering and not part of the IR. If we lowered to python, for example, the secrets module specifies num bits at sample time.

Reply:
Let me add some of this to a cleanup issue